src/gluonnlp/model/bilm_encoder.py

```diff
@@ -0,0 +1,194 @@
+        # pylint: disable=arguments-differ
+        # pylint: disable=unused-argument
+        """Defines the forward computation for cache cell. Arguments can be either
+        :py:class:`NDArray` or :py:class:`Symbol`.
+
+        Parameters
+        ----------
+        inputs : NDArray
+            The input data layout='TNC'.
+        states : Tuple[List[List[NDArray]]]
+            The states. including:
+            states[0] indicates the states used in forward layer,
+            Each layer has a list of two initial tensors with
+            shape (batch_size, proj_size) and (batch_size, hidden_size).
+            states[1] indicates the states used in backward layer,
+            Each layer has a list of two initial tensors with
+            shape (batch_size, proj_size) and (batch_size, hidden_size).
+
+        Returns
+        --------
+        out: NDArray
+            The output data with shape (num_layers, seq_len, batch_size, 2*input_size).
+        [states_forward, states_backward] : List
+            Including:
+            states_forward: The out states from forward layer,
+            which has the same structure with *states[0]*.
+            states_backward: The out states from backward layer,
+            which has the same structure with *states[1]*.
+        """
+        states_forward, states_backward = states
+        if mask is not None:
+            sequence_length = mask.sum(axis=1)
+
+        outputs_forward = []
+        outputs_backward = []
+
+        for layer_index in range(self._num_layers):
+            if layer_index == 0:
+                layer_inputs = inputs
+            else:
+                layer_inputs = outputs_forward[layer_index-1]
+            output, states_forward[layer_index] = F.contrib.foreach(
+                self.forward_layers[layer_index],
+                layer_inputs,
+                states_forward[layer_index])
+            outputs_forward.append(output)
+
+            if layer_index == 0:
+                layer_inputs = F.SequenceReverse(inputs,
+                                                 sequence_length=sequence_length,
+                                                 use_sequence_length=True, axis=0)
+            else:
+                layer_inputs = F.SequenceReverse(outputs_backward[layer_index-1],
+                                                 sequence_length=sequence_length,
+                                                 use_sequence_length=True, axis=0)
+            output, states_backward[layer_index] = F.contrib.foreach(
+                self.backward_layers[layer_index],
+                layer_inputs,
+                states_backward[layer_index])
+            outputs_backward.append(F.SequenceReverse(output,
+                                                      sequence_length=sequence_length,
+                                                      use_sequence_length=True, axis=0))
+        out = F.concat(*[F.stack(*outputs_forward, axis=0),
```

Comment:
I'm not sure if this can be hybridized, as the stack would get a different number of inputs for different lengths.

Reply:
I printed in the code and found that `outputs_forward` is always a list of a single NDArray. This stack should not be used in this case.

Reply:
I see. So the stack is on the layers, which is static.